### PR TITLE
Enable reentrancy for actors

### PR DESCRIPTION
**Pull request by withinboredom**
# Description

This enables actor reentrancy from Dapr 1.2.x

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to
implementation.

Please reference the issue this PR will close: #85 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Tests pass
* [ ] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

**Comment by pkedy** [comment]
@withinboredom What is the status on this?

**Comment by withinboredom** [comment]
It needs documentation and tests but it's here if someone needs it. The biggest blocker is getting the GRPC client done, which is a lot of work.

File: src/lib/DaprClient.php [hunk 1]
```diff
@@ -202,25 +225,15 @@ public function delete(string $url, ?array $params = []): DaprResponse
...
-     *
-     * @throws DaprException
-     */
-    public function shutdown(array $metadata = []): void
+    private function set_current_reentrancy_id(array $headers): array
     {
-        $this->post("/shutdown", $metadata);
+        $current_id = $this->get_current_reentrancy_id();
+
+        return array_merge($headers, ['Dapr-Reentrancy-Id: '.$current_id], $this->extra_headers);
```
Comment (halspang) [inline src/lib/DaprClient.php:232]: If the header already exists, will this create duplicates?

**Review by halspang — COMMENTED** [review]
Thanks for implementing reentrancy! It's been a long time since I've last looked at php, so if you could help walk me through some of this that'd be great. For reentrancy we need:

1. To add the reentrancy field to the config. (This part looks good)
2. Preserve the header on incoming invoke requests.
3. Set the header on outgoing invoke requests from the same call.

It's parts 2 and 3 that I need some clarification on. I can see that we're setting them on the outgoing request and that we have a method to get it from the headers. I'm just not sure exactly how they tie together in the same context. Could you help explain how this works in the php sdk?

File: src/lib/DaprClient.php [hunk 2]
```diff
@@ -114,35 +114,33 @@ private function get_headers(): array
...
         $curl = curl_init($url);
         curl_setopt_array(
             $curl,
             [
                 CURLOPT_RETURNTRANSFER => true,
-                CURLOPT_POST           => true,
-                CURLOPT_HTTPHEADER     => self::as_json(self::get_headers()),
-                CURLOPT_POSTFIELDS     => json_encode($data),
+                CURLOPT_CUSTOMREQUEST  => 'DELETE',
+                CURLOPT_HTTPHEADER     => self::set_current_reentrancy_id(self::as_json(self::get_headers())),
```
[the remark this answers is not in the document]
Comment (withinboredom) [reply src/lib/DaprClient.php:135]: I assume that chains of reentrancy could get arbitrarily long:

```
Actor A --> invoke service B --> invoke service C --> Actor A
```

That may be the wrong thing to do? I'm not sure, but that's why we're passing the header on each type of request.

File: src/lib/DaprClient.php (hunk 1 above)
Comment (withinboredom) [reply src/lib/DaprClient.php:232]: Good catch!

**Comment by withinboredom** [comment]
Closing this and it will be implemented in the new client

**Pull request closed without merging**